cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+                                   StreamSizeSchedule streamSizeSchedule, boolean active,
+                                   long baseRunSize, long baseRunTs, boolean persist) {
+      // TODO add a createScheduleTasks, so that if we create multiple schedules for the same stream at the same
+      // time, we don't have to poll the stream many times
+
+      StreamSizeScheduleTask newTask = new StreamSizeScheduleTask(programId, programType, streamSizeSchedule);
+      StreamSizeScheduleTask previous =
+        scheduleTasks.putIfAbsent(getScheduleId(programId, programType, streamSizeSchedule.getName()),
+                                  newTask);
+      if (previous != null) {
+        // We cannot replace an existing schedule - that functionality is not wanted - yet
+        return;
+      }
+
+      // Initialize the schedule task
+      if (baseRunSize == -1 && baseRunTs == -1) {
+        // This is likely to be the first time that we schedule this task - ie it was not in the schedule store
+        // before. Hence we set the base metrics properly
+        long baseTs = System.currentTimeMillis();
+        long baseSize = pollStream();
+        newTask.startSchedule(baseSize, baseTs, active, persist);
+        synchronized (lastNotificationLock) {
+          lastNotification = new StreamSizeNotification(baseTs, baseSize);
+        }
+      } else {
+        newTask.startSchedule(baseRunSize, baseRunTs, active, persist);
+      }
+
+      if (active) {
+        activeTasks.incrementAndGet();
+      }
+
+      if (lastNotification != null) {
+        // In all cases, when creating a schedule - either if it comes from the store or not,
+        // we want to pass it the last seen notification
+        // This call will send the notification to all the active tasks. The ones which are active before
+        // that method is called will therefore see the notification twice. It is fine though.
+        received(lastNotification, null);
+      }
+    }
+
+    /**
+     * Suspend a scheduling task that is based on the data received by the stream referenced by {@code this} object.
+     */
+    public void suspendScheduleTask(Id.Program programId, SchedulableProgramType programType, String scheduleName) {
+      StreamSizeScheduleTask task = scheduleTasks.get(getScheduleId(programId, programType, scheduleName));
+      if (task == null) {
+        return;
+      }
+      synchronized (this) {
+        if (task.suspend()) {
+          activeTasks.decrementAndGet();
+        }
+      }
+    }
+
+    /**
+     * Resume a scheduling task that is based on the data received by the stream referenced by {@code this} object.
+     */
+    public void resumeScheduleTask(Id.Program programId, SchedulableProgramType programType, String scheduleName) {
+      StreamSizeScheduleTask task = scheduleTasks.get(getScheduleId(programId, programType, scheduleName));
+      if (task == null || !task.resume()) {
```

Comment:
Also, task.resume() must be synchronized or use an Atomic. Otherwise two threads may attempt resume concurrently and both believe they succeeded. Then you would increment the active tasks twice etc.